amd64 compiler bug

**Describe the bug**

While developing [`github.com/ncruces/go-sqlite3`](https://github.com/ncruces/go-sqlite3) I found an issue where a unit test was breaking on my windows machine (but not on GitHub CI).

Basically, running the same code produces a different result on that computer, failing the test. I found that switching to the interpreter fixes the issue on that computer. On other platforms, on GitHub CI, and on a friend's Windows computer, I can't repro the bug. The bug is consistent (always the same output), and survived a computer reboot.

**To Reproduce**

On the affected computer:

```sh
git clone https://github.com/ncruces/wazero-sqlite3-bug.git
cd wazero-sqlite3-bug
go run ./cmd
```

Expected output:
```
[2023/02/08 00:00:00] compiled "2.500000", interpreted "2.500000", want "2.500000"
```

Actual output in affected computer:
```
[2023/02/08 00:00:00] compiled "100000000000000000.000000", interpreted "2.500000", want "2.500000"
```

**Environment (please complete the relevant information):**

- go version go1.20 windows/amd64
- go version go1.19.4 windows/amd64
- wazero Version: v1.0.0-pre.8
- Host architecture: amd64
- Runtime mode: compiler
- Windows: 10.0.19045.2546

**Additional context**

The repo [above](https://github.com/ncruces/wazero-sqlite3-bug.git) runs the test in a [matrix of OS versions](https://github.com/ncruces/wazero-sqlite3-bug/blob/main/.github/workflows/go.yml) (all the ones available on GitHub), and can't trigger the issue.

The [command](https://github.com/ncruces/wazero-sqlite3-bug/blob/main/cmd/main.go) that runs is not actually the failing unit test. It's just an attempt at a simpler case. The tests that fail are [these two](https://github.com/ncruces/wazero-sqlite3-bug/blob/main/stmt_test.go#L203-L208). Also float to text conversions gone bad. I've inspected the code that converts float to text in [SQLite](https://github.com/sqlite/sqlite/blob/master/src/printf.c) and nothing in particular comes up.

Not sure what to make of this…

OK, so running:
```sh
git clone https://github.com/tetratelabs/wazero.git
cd wazero
make test
```

On Git bash, with GNU make installed:
```
ok      github.com/tetratelabs/wazero   4.921s
ok      github.com/tetratelabs/wazero/api       0.623s
?       github.com/tetratelabs/wazero/imports/go        [no test files]
ok      github.com/tetratelabs/wazero/cmd/wazero        45.325s
ok      github.com/tetratelabs/wazero/examples/allocation/rust  0.810s
ok      github.com/tetratelabs/wazero/examples/allocation/tinygo        1.234s
ok      github.com/tetratelabs/wazero/examples/allocation/zig   0.799s
ok      github.com/tetratelabs/wazero/examples/basic    0.722s
ok      github.com/tetratelabs/wazero/examples/cli      13.321s
ok      github.com/tetratelabs/wazero/examples/concurrent-instantiation 1.049s
ok      github.com/tetratelabs/wazero/examples/import-go        0.950s
ok      github.com/tetratelabs/wazero/examples/multiple-results 0.840s
ok      github.com/tetratelabs/wazero/examples/multiple-runtimes        0.951s
ok      github.com/tetratelabs/wazero/experimental      0.704s
ok      github.com/tetratelabs/wazero/experimental/logging      0.570s
ok      github.com/tetratelabs/wazero/imports/assemblyscript    0.571s
ok      github.com/tetratelabs/wazero/imports/assemblyscript/example    0.773s
ok      github.com/tetratelabs/wazero/imports/emscripten        0.567s
?       github.com/tetratelabs/wazero/internal/assemblyscript   [no test files]
ok      github.com/tetratelabs/wazero/imports/go/example        56.367s
--- FAIL: Test_pathReadlink (0.63s)
    require.go:312: expected no error, but was symlink top\top-original-file C:\Users\ncruc\AppData\Local\Temp\Test_pathReadlink1227449904\001/top/top-symlinked: A required privilege is not held by the client.
        C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/fs_test.go:3682
--- FAIL: Test_pathSymlink_errors (0.07s)
    --- FAIL: Test_pathSymlink_errors/success (0.00s)
        require.go:312: expected 0, but was 29: want ESUCCESS but got EIO
            C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/wasi_test.go:148
            C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/fs_test.go:3934
    --- FAIL: Test_pathSymlink_errors/errors (0.00s)
        --- FAIL: Test_pathSymlink_errors/errors/EEXIST (0.00s)
            require.go:312: expected 20, but was 29: want EEXIST but got EIO
                C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/wasi_test.go:148
                C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/fs_test.go:3970
FAIL
FAIL    github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1    16.779s
?       github.com/tetratelabs/wazero/internal/fstest   [no test files]
?       github.com/tetratelabs/wazero/internal/gojs/custom      [no test files]
?       github.com/tetratelabs/wazero/internal/gojs/goos        [no test files]
?       github.com/tetratelabs/wazero/internal/gojs/goarch      [no test files]
?       github.com/tetratelabs/wazero/internal/gojs/run [no test files]
?       github.com/tetratelabs/wazero/internal/ieee754  [no test files]
--- FAIL: Test_main (4.40s)
    --- FAIL: Test_main/zig (2.49s)
        require.go:312: expected "", but was "integer overflowexit_code: 1
            "
            C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/example/cat_test.go:26
--- FAIL: Test_cli (64.79s)
    --- FAIL: Test_cli/zig (6.60s)
        --- FAIL: Test_cli/zig//test.txt (6.60s)
            require.go:312: expected no error, but was exit status 1: integer overflowexit status 1

                C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/example/cat_test.go:98
FAIL
FAIL    github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1/example    70.084s
ok      github.com/tetratelabs/wazero/internal/asm      0.890s
ok      github.com/tetratelabs/wazero/internal/asm/amd64        0.982s
ok      github.com/tetratelabs/wazero/internal/asm/arm64        4.483s
ok      github.com/tetratelabs/wazero/internal/assemblyscript/logging   0.812s
--- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt (0.02s)
    --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz (0.01s)
        --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32 (0.00s)
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32/00000000000000000000000000000000 (0.00s)
                require.go:312: expected 32, but was 0
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:835
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32/00000000000000000000000000000001 (0.00s)
                require.go:312: expected 31, but was 0
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:835
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32/00000000000000000000000000010000 (0.00s)
                require.go:312: expected 27, but was 4
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:835
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32/00000000000000000000000001000000 (0.00s)
                require.go:312: expected 25, but was 6
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:835
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32/10000000000000000000000000000000 (0.00s)
                require.go:312: expected 0, but was 31
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:835
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32/00000000000000000011111111110000 (0.00s)
                require.go:312: expected 18, but was 13
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:835
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32/00000000000000000000000010101010 (0.00s)
                require.go:312: expected 24, but was 7
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:835
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32/00000000000000000001111111111111 (0.00s)
                require.go:312: expected 19, but was 12
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:835
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i32/1111111111111111111111111111111111111111111111111111111111111111 (0.00s)
                require.go:312: expected 0, but was 31
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:835
        --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64 (0.00s)
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64/0000000000000000000000000000000000000000000000000000000000000000 (0.00s)
                require.go:312: expected 64, but was 0
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:837
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64/0000000000000000000000000000000000000000000000000000000000000001 (0.00s)
                require.go:312: expected 63, but was 0
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:837
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64/0000000000000000000000000000000000000000000000000000000000010000 (0.00s)
                require.go:312: expected 59, but was 4
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:837
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64/0000000000000000000000000000000000000000000000000000000001000000 (0.00s)
                require.go:312: expected 57, but was 6
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:837
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64/0000000000000000000000000000000010000000000000000000000000000000 (0.00s)
                require.go:312: expected 32, but was 31
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:837
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64/0000000000000000000000000000000000000000000000000011111111110000 (0.00s)
                require.go:312: expected 50, but was 13
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:837
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64/0000000000000000000000000000000000000000000000000000000010101010 (0.00s)
                require.go:312: expected 56, but was 7
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:837
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64/0000000000000000000000000000000000000000000000000001111111111111 (0.00s)
                require.go:312: expected 51, but was 12
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:837
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Clz/i64/1111111111111111111111111111111111111111111111111111111111111111 (0.00s)
                require.go:312: expected 0, but was 63
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:837
    --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Ctz (0.00s)
        --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Ctz/i32 (0.00s)
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Ctz/i32/00000000000000000000000000000000 (0.00s)
                require.go:312: expected 32, but was 0
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:841
        --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Ctz/i64 (0.00s)
            --- FAIL: TestCompiler_compile_Clz_Ctz_Popcnt/Ctz/i64/0000000000000000000000000000000000000000000000000000000000000000 (0.00s)
                require.go:312: expected 64, but was 0
                    C:/Users/ncruc/Desktop/wazero/internal/engine/compiler/compiler_numeric_test.go:843
FAIL
FAIL    github.com/tetratelabs/wazero/internal/engine/compiler  5.694s
ok      github.com/tetratelabs/wazero/internal/engine/interpreter       0.982s
ok      github.com/tetratelabs/wazero/internal/filecache        0.840s
ok      github.com/tetratelabs/wazero/internal/gojs     0.545s
ok      github.com/tetratelabs/wazero/internal/gojs/logging     1.053s
ok      github.com/tetratelabs/wazero/internal/gojs/util        0.770s
ok      github.com/tetratelabs/wazero/internal/gojs/values      0.879s
ok      github.com/tetratelabs/wazero/internal/integration_test/bench   0.920s
ok      github.com/tetratelabs/wazero/internal/integration_test/engine  0.925s
?       github.com/tetratelabs/wazero/internal/integration_test/vs      [no test files]
--- FAIL: TestSpecTestCompilerCache (16.04s)
    require.go:312: expected no error, but was exit status 1: --- FAIL: TestSpecTestCompilerCache (15.41s)
            --- FAIL: TestSpecTestCompilerCache/i32.wast (0.02s)
                --- FAIL: TestSpecTestCompilerCache/i32.wast/assert_return/line:242 (0.00s)
                    require.go:312: expected true, but was false: i32.wast:242 assert_return invoke clz ([{type: i32, value: 4294967295}])
                                have [31]
                                want [0]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i32.wast/assert_return/line:243 (0.00s)
                    require.go:312: expected true, but was false: i32.wast:243 assert_return invoke clz ([{type: i32, value: 0}])
                                have [0]
                                want [32]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i32.wast/assert_return/line:244 (0.00s)
                    require.go:312: expected true, but was false: i32.wast:244 assert_return invoke clz ([{type: i32, value: 32768}])
                                have [15]
                                want [16]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i32.wast/assert_return/line:245 (0.00s)
                    require.go:312: expected true, but was false: i32.wast:245 assert_return invoke clz ([{type: i32, value: 255}])
                                have [7]
                                want [24]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i32.wast/assert_return/line:246 (0.00s)
                    require.go:312: expected true, but was false: i32.wast:246 assert_return invoke clz ([{type: i32, value: 2147483648}])
                                have [31]
                                want [0]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i32.wast/assert_return/line:247 (0.00s)
                    require.go:312: expected true, but was false: i32.wast:247 assert_return invoke clz ([{type: i32, value: 1}])
                                have [0]
                                want [31]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i32.wast/assert_return/line:248 (0.00s)
                    require.go:312: expected true, but was false: i32.wast:248 assert_return invoke clz ([{type: i32, value: 2}])
                                have [1]
                                want [30]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i32.wast/assert_return/line:249 (0.00s)
                    require.go:312: expected true, but was false: i32.wast:249 assert_return invoke clz ([{type: i32, value: 2147483647}])
                                have [30]
                                want [1]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i32.wast/assert_return/line:252 (0.00s)
                    require.go:312: expected true, but was false: i32.wast:252 assert_return invoke ctz ([{type: i32, value: 0}])
                                have [0]
                                want [32]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
            --- FAIL: TestSpecTestCompilerCache/i64.wast (0.03s)
                --- FAIL: TestSpecTestCompilerCache/i64.wast/assert_return/line:242 (0.00s)
                    require.go:312: expected true, but was false: i64.wast:242 assert_return invoke clz ([{type: i64, value: 18446744073709551615}])
                                have [63]
                                want [0]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i64.wast/assert_return/line:243 (0.00s)
                    require.go:312: expected true, but was false: i64.wast:243 assert_return invoke clz ([{type: i64, value: 0}])
                                have [0]
                                want [64]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i64.wast/assert_return/line:244 (0.00s)
                    require.go:312: expected true, but was false: i64.wast:244 assert_return invoke clz ([{type: i64, value: 32768}])
                                have [15]
                                want [48]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i64.wast/assert_return/line:245 (0.00s)
                    require.go:312: expected true, but was false: i64.wast:245 assert_return invoke clz ([{type: i64, value: 255}])
                                have [7]
                                want [56]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i64.wast/assert_return/line:246 (0.00s)
                    require.go:312: expected true, but was false: i64.wast:246 assert_return invoke clz ([{type: i64, value: 9223372036854775808}])
                                have [63]
                                want [0]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i64.wast/assert_return/line:247 (0.00s)
                    require.go:312: expected true, but was false: i64.wast:247 assert_return invoke clz ([{type: i64, value: 1}])
                                have [0]
                                want [63]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i64.wast/assert_return/line:248 (0.00s)
                    require.go:312: expected true, but was false: i64.wast:248 assert_return invoke clz ([{type: i64, value: 2}])
                                have [1]
                                want [62]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i64.wast/assert_return/line:249 (0.00s)
                    require.go:312: expected true, but was false: i64.wast:249 assert_return invoke clz ([{type: i64, value: 9223372036854775807}])
                                have [62]
                                want [1]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
                --- FAIL: TestSpecTestCompilerCache/i64.wast/assert_return/line:252 (0.00s)
                    require.go:312: expected true, but was false: i64.wast:252 assert_return invoke ctz ([{type: i64, value: 0}])
                                have [0]
                                want [64]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
            --- FAIL: TestSpecTestCompilerCache/load.wast (0.01s)
                --- FAIL: TestSpecTestCompilerCache/load.wast/assert_return/line:201 (0.00s)
                    require.go:312: expected true, but was false: load.wast:201 assert_return invoke as-unary-operand ([])
                                have [0]
                                want [32]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
            --- FAIL: TestSpecTestCompilerCache/memory_grow.wast (0.09s)
                --- FAIL: TestSpecTestCompilerCache/memory_grow.wast/assert_return/line:299 (0.00s)
                    require.go:312: expected true, but was false: memory_grow.wast:299 assert_return invoke as-unary-operand ([])
                                have [0]
                                want [31]
                        C:/Users/ncruc/Desktop/wazero/internal/integration_test/spectest/spectest.go:439
        FAIL

        C:/Users/ncruc/Desktop/wazero/internal/integration_test/filecache/filecache_test.go:50
FAIL
FAIL    github.com/tetratelabs/wazero/internal/integration_test/filecache       17.837s
ok      github.com/tetratelabs/wazero/internal/integration_test/fs      0.794s
ok      github.com/tetratelabs/wazero/internal/integration_test/fuzzcases       9.035s
ok      github.com/tetratelabs/wazero/internal/integration_test/spectest        0.711s
ok      github.com/tetratelabs/wazero/internal/integration_test/vs/compiler     2.298s
ok      github.com/tetratelabs/wazero/internal/integration_test/vs/interpreter  12.228s
?       github.com/tetratelabs/wazero/internal/testing/dwarftestdata    [no test files]
?       github.com/tetratelabs/wazero/internal/testing/enginetest       [no test files]
ok      github.com/tetratelabs/wazero/internal/leb128   14.659s
ok      github.com/tetratelabs/wazero/internal/logging  0.729s
ok      github.com/tetratelabs/wazero/internal/moremath 0.932s
ok      github.com/tetratelabs/wazero/internal/platform 0.909s
ok      github.com/tetratelabs/wazero/internal/sys      0.878s
--- FAIL: TestDirFS_Readlink (0.02s)
    require.go:312: expected no error, but was symlink animals.txt C:\Users\ncruc\AppData\Local\Temp\TestDirFS_Readlink3161770538\001/symlinked-animals.txt: A required privilege is not held by the client.
        C:/Users/ncruc/Desktop/wazero/internal/sysfs/dirfs_test.go:581
--- FAIL: TestDirFS_Symlink (0.09s)
    require.go:312: expected errors.Is(A required privilege is not held by the client., file exists), but it wasn't
        C:/Users/ncruc/Desktop/wazero/internal/sysfs/dirfs_test.go:543
FAIL
FAIL    github.com/tetratelabs/wazero/internal/sysfs    1.200s
?       github.com/tetratelabs/wazero/internal/testing/hammer   [no test files]
?       github.com/tetratelabs/wazero/internal/testing/maintester       [no test files]
?       github.com/tetratelabs/wazero/internal/testing/proxy    [no test files]
ok      github.com/tetratelabs/wazero/internal/testing/fs       0.866s
?       github.com/tetratelabs/wazero/internal/version  [no test files]
ok      github.com/tetratelabs/wazero/internal/testing/require  0.764s
ok      github.com/tetratelabs/wazero/internal/u32      0.693s
ok      github.com/tetratelabs/wazero/internal/u64      0.823s
ok      github.com/tetratelabs/wazero/internal/wasi_snapshot_preview1   0.825s
ok      github.com/tetratelabs/wazero/internal/wasi_snapshot_preview1/logging   1.115s
ok      github.com/tetratelabs/wazero/internal/wasm     1.569s
ok      github.com/tetratelabs/wazero/internal/wasm/binary      0.867s
?       github.com/tetratelabs/wazero/internal/wasmruntime      [no test files]
ok      github.com/tetratelabs/wazero/internal/wasmdebug        2.055s
ok      github.com/tetratelabs/wazero/internal/wazeroir 0.474s
ok      github.com/tetratelabs/wazero/sys       0.421s
FAIL
make: *** [test] Error 1
```

Multiple failures, but the big one seems to be related to `Clz_Ctz_Popcnt`.

PS: The symlink one may affect others. This might have changed recently, but symlinks used to require admin privileges on Windows, so it's expected that they may fail.

Yes, that fixes the issue. The only failing tests after it are symlink ones:
```
ok      github.com/tetratelabs/wazero   3.255s
ok      github.com/tetratelabs/wazero/api       (cached)
?       github.com/tetratelabs/wazero/imports/go        [no test files]
ok      github.com/tetratelabs/wazero/cmd/wazero        17.422s
ok      github.com/tetratelabs/wazero/examples/allocation/rust  0.679s
ok      github.com/tetratelabs/wazero/examples/allocation/tinygo        0.668s
ok      github.com/tetratelabs/wazero/examples/allocation/zig   0.601s
ok      github.com/tetratelabs/wazero/examples/basic    0.681s
ok      github.com/tetratelabs/wazero/examples/cli      (cached)
ok      github.com/tetratelabs/wazero/examples/concurrent-instantiation 0.685s
ok      github.com/tetratelabs/wazero/examples/import-go        0.666s
ok      github.com/tetratelabs/wazero/examples/multiple-results 0.663s
ok      github.com/tetratelabs/wazero/examples/multiple-runtimes        0.703s
ok      github.com/tetratelabs/wazero/experimental      0.626s
ok      github.com/tetratelabs/wazero/experimental/logging      0.302s
ok      github.com/tetratelabs/wazero/imports/assemblyscript    0.390s
ok      github.com/tetratelabs/wazero/imports/assemblyscript/example    0.570s
ok      github.com/tetratelabs/wazero/imports/emscripten        0.419s
?       github.com/tetratelabs/wazero/internal/assemblyscript   [no test files]
?       github.com/tetratelabs/wazero/internal/fstest   [no test files]
ok      github.com/tetratelabs/wazero/imports/go/example        19.205s
--- FAIL: Test_pathReadlink (0.02s)
    require.go:312: expected no error, but was symlink top\top-original-file C:\Users\ncruc\AppData\Local\Temp\Test_pathReadlink2953894191\001/top/top-symlinked: A required privilege is not held by the client.
        C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/fs_test.go:3682
--- FAIL: Test_pathSymlink_errors (0.03s)
    --- FAIL: Test_pathSymlink_errors/success (0.00s)
        require.go:312: expected 0, but was 29: want ESUCCESS but got EIO
            C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/wasi_test.go:148
            C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/fs_test.go:3934
    --- FAIL: Test_pathSymlink_errors/errors (0.00s)
        --- FAIL: Test_pathSymlink_errors/errors/EEXIST (0.00s)
            require.go:312: expected 20, but was 29: want EEXIST but got EIO
                C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/wasi_test.go:148
                C:/Users/ncruc/Desktop/wazero/imports/wasi_snapshot_preview1/fs_test.go:3970
FAIL
FAIL    github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1    5.817s
?       github.com/tetratelabs/wazero/internal/gojs/custom      [no test files]
?       github.com/tetratelabs/wazero/internal/gojs/goarch      [no test files]
?       github.com/tetratelabs/wazero/internal/gojs/goos        [no test files]
?       github.com/tetratelabs/wazero/internal/gojs/run [no test files]
?       github.com/tetratelabs/wazero/internal/ieee754  [no test files]
?       github.com/tetratelabs/wazero/internal/integration_test/vs      [no test files]
?       github.com/tetratelabs/wazero/internal/testing/dwarftestdata    [no test files]
?       github.com/tetratelabs/wazero/internal/testing/enginetest       [no test files]
?       github.com/tetratelabs/wazero/internal/testing/hammer   [no test files]
?       github.com/tetratelabs/wazero/internal/testing/maintester       [no test files]
?       github.com/tetratelabs/wazero/internal/testing/proxy    [no test files]
?       github.com/tetratelabs/wazero/internal/version  [no test files]
?       github.com/tetratelabs/wazero/internal/wasmruntime      [no test files]
ok      github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1/example    52.943s
ok      github.com/tetratelabs/wazero/internal/asm      (cached)
ok      github.com/tetratelabs/wazero/internal/asm/amd64        (cached)
ok      github.com/tetratelabs/wazero/internal/asm/arm64        (cached)
ok      github.com/tetratelabs/wazero/internal/assemblyscript/logging   (cached)
ok      github.com/tetratelabs/wazero/internal/engine/compiler  5.717s
ok      github.com/tetratelabs/wazero/internal/engine/interpreter       (cached)
ok      github.com/tetratelabs/wazero/internal/filecache        (cached)
ok      github.com/tetratelabs/wazero/internal/gojs     0.501s
ok      github.com/tetratelabs/wazero/internal/gojs/logging     (cached)
ok      github.com/tetratelabs/wazero/internal/gojs/util        (cached)
ok      github.com/tetratelabs/wazero/internal/gojs/values      (cached)
ok      github.com/tetratelabs/wazero/internal/integration_test/bench   0.752s
ok      github.com/tetratelabs/wazero/internal/integration_test/engine  0.533s
ok      github.com/tetratelabs/wazero/internal/integration_test/filecache       35.649s
ok      github.com/tetratelabs/wazero/internal/integration_test/fs      0.903s
ok      github.com/tetratelabs/wazero/internal/integration_test/fuzzcases       7.662s
ok      github.com/tetratelabs/wazero/internal/integration_test/spectest        (cached)
ok      github.com/tetratelabs/wazero/internal/integration_test/vs/compiler     1.829s
ok      github.com/tetratelabs/wazero/internal/integration_test/vs/interpreter  11.517s
ok      github.com/tetratelabs/wazero/internal/leb128   (cached)
ok      github.com/tetratelabs/wazero/internal/logging  (cached)
ok      github.com/tetratelabs/wazero/internal/moremath (cached)
ok      github.com/tetratelabs/wazero/internal/platform (cached)
ok      github.com/tetratelabs/wazero/internal/sys      (cached)
--- FAIL: TestDirFS_Readlink (0.03s)
    require.go:312: expected no error, but was symlink animals.txt C:\Users\ncruc\AppData\Local\Temp\TestDirFS_Readlink2319260500\001/symlinked-animals.txt: A required privilege is not held by the client.
        C:/Users/ncruc/Desktop/wazero/internal/sysfs/dirfs_test.go:581
--- FAIL: TestDirFS_Symlink (0.22s)
    require.go:312: expected errors.Is(A required privilege is not held by the client., file exists), but it wasn't
        C:/Users/ncruc/Desktop/wazero/internal/sysfs/dirfs_test.go:543
FAIL
FAIL    github.com/tetratelabs/wazero/internal/sysfs    1.612s
ok      github.com/tetratelabs/wazero/internal/testing/fs       (cached)
ok      github.com/tetratelabs/wazero/internal/testing/require  (cached)
ok      github.com/tetratelabs/wazero/internal/u32      (cached)
ok      github.com/tetratelabs/wazero/internal/u64      (cached)
ok      github.com/tetratelabs/wazero/internal/wasi_snapshot_preview1   (cached)
ok      github.com/tetratelabs/wazero/internal/wasi_snapshot_preview1/logging   (cached)
ok      github.com/tetratelabs/wazero/internal/wasm     (cached)
ok      github.com/tetratelabs/wazero/internal/wasm/binary      (cached)
ok      github.com/tetratelabs/wazero/internal/wasmdebug        (cached)
ok      github.com/tetratelabs/wazero/internal/wazeroir (cached)
ok      github.com/tetratelabs/wazero/sys       (cached)
FAIL
make: *** [test] Error 1
```

And my use case works as well:
```
2023/02/09 12:00:00 compiled "2.500000", interpreted "2.500000", want "2.500000"
```

But singling out windows is maybe not the correct fix here? Since building and running on WSL gives the old behaviour:
```
2023/02/09 12:00:00 compiled "100000000000000000.000000", interpreted "2.500000", want "2.500000"
```

My CPU is missing `LZCNT`, should that be tested at runtime?
I also think it's fair to simply disable the compiler in such cases, if that makes it easier.

--- a/internal/engine/compiler/impl_amd64.go
+++ b/internal/engine/compiler/impl_amd64.go
@@ -8,7 +8,6 @@ import (
 	"bytes"
 	"fmt"
 	"math"
-	"runtime"
 
 	"github.com/tetratelabs/wazero/internal/asm"
 	"github.com/tetratelabs/wazero/internal/asm/amd64"
@@ -83,8 +82,9 @@ func (c *amd64Compiler) compileNOP() asm.Node {
 }
 
 type amd64Compiler struct {
-	assembler amd64.Assembler
-	ir        *wazeroir.CompilationResult
+	assembler   amd64.Assembler
+	ir          *wazeroir.CompilationResult
+	cpuFeatures platform.CpuFeatureFlags
 	// locationStack holds the state of wazeroir virtual stack.
 	// and each item is either placed in register or the actual memory stack.
 	locationStack *runtimeValueLocationStack
@@ -103,6 +103,7 @@ func newAmd64Compiler() compiler {
 	c := &amd64Compiler{
 		assembler:     amd64.NewAssembler(),
 		locationStack: newRuntimeValueLocationStack(),
+		cpuFeatures:   platform.CpuFeatures,
 	}
 	return c
 }
@@ -114,6 +115,7 @@ func (c *amd64Compiler) Init(ir *wazeroir.CompilationResult, withListener bool) 
 	*c = amd64Compiler{
 		labels:       map[string]*amd64LabelInfo{},
 		ir:           ir,
+		cpuFeatures:  c.cpuFeatures,
 		withListener: withListener,
 		currentLabel: wazeroir.EntrypointLabel,
 	}
@@ -1170,7 +1172,7 @@ func (c *amd64Compiler) compileClz(o *wazeroir.OperationClz) error {
 		return err
 	}
 
-	if runtime.GOOS != "darwin" && runtime.GOOS != "freebsd" {
+	if c.cpuFeatures.HasExtra(platform.CpuExtraFeatureABM) {
 		if o.Type == wazeroir.UnsignedInt32 {
 			c.assembler.CompileRegisterToRegister(amd64.LZCNTL, target.register, target.register)
 		} else {
@@ -1233,7 +1235,7 @@ func (c *amd64Compiler) compileCtz(o *wazeroir.OperationCtz) error {
 		return err
 	}
 
-	if runtime.GOOS != "darwin" && runtime.GOOS != "freebsd" {
+	if c.cpuFeatures.HasExtra(platform.CpuExtraFeatureABM) {
 		if o.Type == wazeroir.UnsignedInt32 {
 			c.assembler.CompileRegisterToRegister(amd64.TZCNTL, target.register, target.register)
 		} else {

--- a/internal/engine/compiler/impl_amd64_test.go
+++ b/internal/engine/compiler/impl_amd64_test.go
@@ -1,11 +1,13 @@
 package compiler
 
 import (
+	"encoding/hex"
 	"testing"
 	"unsafe"
 
 	"github.com/tetratelabs/wazero/internal/asm"
 	"github.com/tetratelabs/wazero/internal/asm/amd64"
+	"github.com/tetratelabs/wazero/internal/platform"
 	"github.com/tetratelabs/wazero/internal/testing/require"
 	"github.com/tetratelabs/wazero/internal/wasm"
 	"github.com/tetratelabs/wazero/internal/wazeroir"
@@ -438,6 +440,132 @@ func TestAmd64Compiler_preventCrossedTargetdRegisters(t *testing.T) {
 		restoreCrossing()
 		// Require initial state after restoring.
 		require.Equal(t, initialRegisters, collectRegistersFromRuntimeValues(tt.initial))
+	}
+}
+
+// mockCpuFlags implements platform.CpuFeatureFlags
+type mockCpuFlags struct {
+	flags      uint64
+	extraFlags uint64
+}
+
+// Has implements the method of the same name in platform.CpuFeatureFlags
+func (f *mockCpuFlags) Has(flag uint64) bool {
+	return (f.flags & flag) != 0
+}
+
+// HasExtra implements the method of the same name in platform.CpuFeatureFlags
+func (f *mockCpuFlags) HasExtra(flag uint64) bool {
+	return (f.extraFlags & flag) != 0
+}
+
+// Relates to #1111 (Clz): older AMD64 CPUs do not support the LZCNT instruction
+// CPUID should be used instead. We simulate presence/absence of the feature
+// by overriding the field in the corresponding struct.
+func TestAmd64Compiler_ensureClz_ABM(t *testing.T) {
+	tests := []struct {
+		name         string
+		cpuFeatures  platform.CpuFeatureFlags
+		expectedCode string
+	}{
+		{
+			name:         "with ABM",
+			expectedCode: "b80a000000f3480fbdc0",
+			cpuFeatures: &mockCpuFlags{
+				flags:      0,
+				extraFlags: platform.CpuExtraFeatureABM,
+			},
+		},
+		{
+			name:         "without ABM",
+			expectedCode: "b80a0000004883f8007507b840000000eb08480fbdc04883f03f",
+			cpuFeatures: &mockCpuFlags{
+				flags:      0,
+				extraFlags: 0, // no flags, thus no ABM, i.e. no LZCNT
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			env := newCompilerEnvironment()
+
+			newCompiler := func() compiler {
+				c := newCompiler().(*amd64Compiler)
+				// override auto-detected CPU features with the test case
+				c.cpuFeatures = tt.cpuFeatures
+				return c
+			}
+
+			compiler := env.requireNewCompiler(t, newCompiler, nil)
+
+			err := compiler.compileConstI32(&wazeroir.OperationConstI32{Value: 10})
+			require.NoError(t, err)
+
+			err = compiler.compileClz(&wazeroir.OperationClz{Type: wazeroir.UnsignedInt64})
+			require.NoError(t, err)
+
+			compiler.compileNOP() // pad for jump target (when no ABM)
+
+			code, _, err := compiler.compile()
+			require.NoError(t, err)
+
+			require.Equal(t, tt.expectedCode, hex.EncodeToString(code))
+		})
+	}
+}
+
+// Relates to #1111 (Ctz): older AMD64 CPUs do not support the LZCNT instruction
+// CPUID should be used instead. We simulate presence/absence of the feature
+// by overriding the field in the corresponding struct.
+func TestAmd64Compiler_ensureCtz_ABM(t *testing.T) {
+	tests := []struct {
+		name         string
+		cpuFeatures  platform.CpuFeatureFlags
+		expectedCode string
+	}{
+		{
+			name:         "with ABM",
+			expectedCode: "b80a000000f3480fbcc0",
+			cpuFeatures: &mockCpuFlags{
+				flags:      0,
+				extraFlags: platform.CpuExtraFeatureABM,
+			},
+		},
+		{
+			name:         "without ABM",
+			expectedCode: "b80a0000004883f8007507b840000000eb05f3480fbcc0",
+			cpuFeatures: &mockCpuFlags{
+				flags:      0,
+				extraFlags: 0, // no flags, thus no ABM, i.e. no LZCNT
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			env := newCompilerEnvironment()
+
+			newCompiler := func() compiler {
+				c := newCompiler().(*amd64Compiler)
+				// override auto-detected CPU features with the test case
+				c.cpuFeatures = tt.cpuFeatures
+				return c
+			}
+
+			compiler := env.requireNewCompiler(t, newCompiler, nil)
+
+			err := compiler.compileConstI32(&wazeroir.OperationConstI32{Value: 10})
+			require.NoError(t, err)
+
+			err = compiler.compileCtz(&wazeroir.OperationCtz{Type: wazeroir.UnsignedInt64})
+			require.NoError(t, err)
+
+			compiler.compileNOP() // pad for jump target (when no ABM)
+
+			code, _, err := compiler.compile()
+			require.NoError(t, err)
+
+			require.Equal(t, tt.expectedCode, hex.EncodeToString(code))
+		})
 	}
 }
 

--- a/internal/platform/cpuid_amd64.go
+++ b/internal/platform/cpuid_amd64.go
@@ -1,0 +1,79 @@
+package platform
+
+const (
+	// CpuFeatureSSE3 is the flag to query CpuFeatureFlags.Has for SSEv3 capabilities
+	CpuFeatureSSE3 = uint64(1)
+	// CpuFeatureSSE4_1 is the flag to query CpuFeatureFlags.Has for SSEv4.1 capabilities
+	CpuFeatureSSE4_1 = uint64(1) << 19
+	// CpuFeatureSSE4_2 is the flag to query CpuFeatureFlags.Has for SSEv4.2 capabilities
+	CpuFeatureSSE4_2 = uint64(1) << 20
+)
+
+const (
+	// CpuExtraFeatureABM is the flag to query CpuFeatureFlags.HasExtra for Advanced Bit Manipulation capabilities (e.g. LZCNT)
+	CpuExtraFeatureABM = uint64(1) << 5
+)
+
+// CpuFeatures exposes the capabilities for this CPU, queried via the Has, HasExtra methods
+var CpuFeatures CpuFeatureFlags = loadCpuFeatureFlags()
+
+// CpuFeatureFlags exposes methods for querying CPU capabilities
+type CpuFeatureFlags interface {
+	// Has returns true when the specified flag (represented as uint64) is supported
+	Has(cpuFeature uint64) bool
+	// HasExtra returns true when the specified extraFlag (represented as uint64) is supported
+	HasExtra(cpuFeature uint64) bool
+}
+
+// cpuFeatureFlags implements CpuFeatureFlags interface
+type cpuFeatureFlags struct {
+	flags      uint64
+	extraFlags uint64
+}
+
+// cpuid exposes the CPUID instruction to the Go layer (https://www.amd.com/system/files/TechDocs/25481.pdf)
+// implemented in impl_amd64.s
+func cpuid(arg1, arg2 uint32) (eax, ebx, ecx, edx uint32)
+
+// cpuidAsBitmap combines the result of invoking cpuid to uint64 bitmap
+func cpuidAsBitmap(arg1, arg2 uint32) uint64 {
+	_ /* eax */, _ /* ebx */, ecx, edx := cpuid(arg1, arg2)
+	return (uint64(edx) << 32) | uint64(ecx)
+}
+
+// loadStandardRange load flags from the standard range, panics otherwise
+func loadStandardRange(id uint32) uint64 {
+	// ensure that the id is in the valid range, returned by cpuid(0,0)
+	maxRange, _, _, _ := cpuid(0, 0)
+	if id > maxRange {
+		panic("cannot query standard CPU flags")
+	}
+	return cpuidAsBitmap(id, 0)
+}
+
+// loadStandardRange load flags from the extended range, panics otherwise
+func loadExtendedRange(id uint32) uint64 {
+	// ensure that the id is in the valid range, returned by cpuid(0x80000000,0)
+	maxRange, _, _, _ := cpuid(0x80000000, 0)
+	if id > maxRange {
+		panic("cannot query extended CPU flags")
+	}
+	return cpuidAsBitmap(id, 0)
+}
+
+func loadCpuFeatureFlags() CpuFeatureFlags {
+	return &cpuFeatureFlags{
+		flags:      loadStandardRange(1),
+		extraFlags: loadExtendedRange(0x80000001),
+	}
+}
+
+// Has implements the same method on the CpuFeatureFlags interface
+func (f *cpuFeatureFlags) Has(cpuFeature uint64) bool {
+	return (f.flags & cpuFeature) != 0
+}
+
+// HasExtra implements the same method on the CpuFeatureFlags interface
+func (f *cpuFeatureFlags) HasExtra(cpuFeature uint64) bool {
+	return (f.extraFlags & cpuFeature) != 0
+}

--- a/internal/platform/cpuid_amd64.s
+++ b/internal/platform/cpuid_amd64.s
@@ -1,0 +1,14 @@
+#include "textflag.h"
+
+// lifted from github.com/intel-go/cpuid and src/internal/cpu/cpu_x86.s
+// func cpuid(arg1, arg2 uint32) (eax, ebx, ecx, edx uint32)
+TEXT ·cpuid(SB), NOSPLIT, $0-24
+	MOVL arg1+0(FP), AX
+	MOVL arg2+4(FP), CX
+	CPUID
+	MOVL AX, eax+8(FP)
+	MOVL BX, ebx+12(FP)
+	MOVL CX, ecx+16(FP)
+	MOVL DX, edx+20(FP)
+	RET
+

--- a/internal/platform/cpuid_amd64_test.go
+++ b/internal/platform/cpuid_amd64_test.go
@@ -1,0 +1,18 @@
+package platform
+
+import (
+	"testing"
+
+	"github.com/tetratelabs/wazero/internal/testing/require"
+)
+
+func TestAmd64CpuId_cpuHasFeature(t *testing.T) {
+	flags := cpuFeatureFlags{
+		flags:      CpuFeatureSSE3,
+		extraFlags: CpuExtraFeatureABM,
+	}
+	require.True(t, flags.Has(CpuFeatureSSE3))
+	require.False(t, flags.Has(CpuFeatureSSE4_2))
+	require.True(t, flags.HasExtra(CpuExtraFeatureABM))
+	require.False(t, flags.HasExtra(1<<6)) // some other value
+}

